pkg/bgp/controller/k8s.go

```diff
@@ -0,0 +1,45 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+
+	"github.com/cilium/cilium/pkg/k8s"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// client wraps the K8s client so that it can conform to the client from
+// MetalLB.
+type client struct {
+	*k8s.K8sClient
+}
+
+func (c *client) Update(svc *v1.Service) (*v1.Service, error) {
+	return c.CoreV1().Services(svc.GetNamespace()).Update(context.TODO(), svc, metav1.UpdateOptions{})
+}
+func (c *client) UpdateStatus(svc *v1.Service) error {
+	_, err := c.CoreV1().Services(svc.GetNamespace()).UpdateStatus(context.TODO(), svc, metav1.UpdateOptions{})
```

Comment:
Yeah, ideally we should not be using `context.TODO()` anywhere in Cilium, and especially while in communication with the apiserver. We have an open issue to track this (https://github.com/cilium/cilium/issues/10655), but haven't had the time to go through it. (On me as I'm assigned to it :sweat_smile:)

FWIW, this hasn't caused us problems in the past, but that doesn't preclude it from happening in the future. I'll take a note to followup on this, but this time specifically focus on the BGP subsystem instead of tackling the aforementioned issue as a whole. The reason it's a followup is because I don't think it's as simple as just slapping on a 2m timeout (for ex); we will need to think about the error flow and what actually happens when the context deadlines, etc.

Reply:
> this hasn't caused us problems in the past

[citation needed] ;)

Main reason I bring this up is that we had a whole bunch of trouble with the etcd clients when testing high availability / failover scenarios and it took quite a bit of work to get those codepaths into a decent state. Beyond that, if we just keep adding `context.TODO()` into the codebase then we're going backwards on progress to #10655.

I would consider this a requirement before marking this feature as stable but I agree it needs more careful thought & consideration.